mackerel-plugin-gcp-compute-engine/lib/gcp-compute-engine.go

```diff
@@ -0,0 +1,202 @@
+		Unit:  "float",
+		Metrics: []mp.Metrics{
+			{Name: "read_ops_count", Label: "Read Ops Count"},
+			{Name: "write_ops_count", Label: "Write Ops Count"},
+		},
+	},
+	"Network.BytesCount": mp.Graphs{
+		Label: "Network Received Bytes Count",
+		Unit:  "bytes",
+		Metrics: []mp.Metrics{
+			{Name: "received_bytes_count", Label: "Received Bytes Count"},
+			{Name: "sent_bytes_count", Label: "Sent Bytes Count"},
+		},
+	},
+	"Network.PacketsCount": mp.Graphs{
+		Label: "Network Received Packets Count",
+		Unit:  "float",
+		Metrics: []mp.Metrics{
+			{Name: "received_packets_count", Label: "Received Packets Count"},
+			{Name: "sent_packets_count", Label: "Sent Packets Count"},
+		},
+	},
+}
+
+// GraphDefinition is return graphdef
+func (p ComputeEnginePlugin) GraphDefinition() map[string]mp.Graphs {
+	return graphdef
+}
+
+func getLatestValue(listCall *monitoring.ProjectsTimeSeriesListCall, filter string, startTime string, endTime string, opts *Option) (float64, error) {
+	res, err := listCall.Filter(filter).IntervalEndTime(endTime).IntervalStartTime(startTime).Do(*opts)
+
+	if err != nil || res == nil {
+		return 0, err
+	}
+
+	valuePtr := res.TimeSeries[0].Points[0].Value
+
+	var value float64
+	if valuePtr.Int64Value != nil {
+		value = float64(*valuePtr.Int64Value)
+	} else if valuePtr.DoubleValue != nil {
+		value = *valuePtr.DoubleValue
+	}
+
+	return value, nil
+}
+
+func mkFilter(domain string, metricName string, instance string) string {
+	filter := `metric.type = "` + domain + metricName + `"`
+	switch domain {
+	case computeDomain:
+		filter += " AND metric.label.instance_name = " + instance
+	}
+
+	return filter
+}
+
+// FetchMetrics fetches metrics from Google Monitoring API
+func (p ComputeEnginePlugin) FetchMetrics() (map[string]float64, error) {
+	now := time.Now()
+	formattedEnd := now.Format(zuluFormat)
+	m, _ := time.ParseDuration(duration)
+	formattedStart := now.Add(-m).Format(zuluFormat)
+	listCall := p.MonitoringService.Projects.TimeSeries.List(p.Project)
+
+	stat := map[string]float64{}
+	for _, metricName := range []string{
+		"/firewall/dropped_bytes_count",
+		"/firewall/dropped_packets_count",
+		"/instance/cpu/utilization",
+		"/instance/disk/read_bytes_count",
+		"/instance/disk/read_ops_count",
+		"/instance/disk/write_bytes_count",
+		"/instance/disk/write_ops_count",
+		"/instance/network/received_bytes_count",
+		"/instance/network/received_packets_count",
+		"/instance/network/sent_bytes_count",
+		"/instance/network/sent_packets_count",
+	} {
+		value, err := getLatestValue(listCall, mkFilter(computeDomain, metricName, p.InstanceName), formattedStart, formattedEnd, p.Option)
+		if err != nil {
+			continue
+		}
+		splited := strings.Split(metricName, "/")
+		stat[splited[len(splited) - 1]] = value
+	}
+
+	return stat, nil
+}
+
+// Do the plugin
+func Do() {
+	optProject := flag.String("project", "", "Project No")
+	optInstanceName := flag.String("instance-name", "", "Instance Name")
```

Comment:
I checked and these auto detections seems to work properly, but I think following behavior is better (Sorry that I didn't write the previous comment so...)

1. If command line parameters are specified, use the value
2. Otherwise Try to get from metadata API